txpool/src/containers/dependency.rs

```diff
@@ -0,0 +1,482 @@
+                                    )?;
+                                } else {
+                                    // tx output is in pool
+                                    let output_tx = txs.get(utxo_id.tx_id()).unwrap();
+                                    let output =
+                                        &output_tx.outputs()[utxo_id.output_index() as usize];
+                                    Self::check_if_coin_input_can_spend_output(
+                                        output, input, false,
+                                    )?;
+                                };
+
+                                collided.push(*spend_by);
+                            }
+                        }
+                        // if coin is not spend, it will be spend later down the line
+                    } else {
+                        // fetch from db and check if tx exist.
+                        if let Some(db_tx) = db.transaction(*utxo_id.tx_id()) {
+                            if let Some(output) =
+                                db_tx.outputs().get(utxo_id.output_index() as usize)
+                            {
+                                // add depth
+                                max_depth = core::cmp::max(1, max_depth);
+                                // do all checks that we can do
+                                Self::check_if_coin_input_can_spend_output(output, input, true)?;
+                                // insert it as spend coin.
+                                // check for double spend should be done before transaction is received.
+                                db_coins.insert(
+                                    *utxo_id,
+                                    CoinState {
+                                        is_spend_by: Some(tx.id() as TxId),
+                                        depth: 0,
+                                    },
+                                );
+                            } else {
+                                // output out of bound
+                                return Err(Error::NotInsertedOutputNotExisting(*utxo_id).into());
+                            }
+                        } else {
+                            return Err(
+                                Error::NotInsertedInputTxNotExisting(*utxo_id.tx_id()).into()
+                            );
+                        }
+                    }
+
+                    // yey we got our coin
+                }
+                Input::Contract { contract_id, .. } => {
+                    // Does contract exist. We dont need to do any check here other then if contract_id exist or not.
+                    if let Some(state) = self.contracts.get(contract_id) {
+                        // check if contract is created after this transaction.
+                        if state.gas_price > tx.gas_price() {
+                            return Err(Error::NotInsertedContractPricedLower(*contract_id).into());
```

Comment:
Seems like we should have a test case to ensure this error is triggered correctly

Reply:
There was a bug here it should be `if tx.gas_price() > state.gas_price { `, good catch. Added tests to check for it.